touchables/TouchableWithoutFeedback.js

```diff
@@ -0,0 +1,346 @@
+        } else if (this.pressInTimeout) {
+          clearTimeout(this.pressInTimeout);
+          this.pressInTimeout = null;
+          this.moveToState(TOUCHABLE_STATE.BEGAN);
+        }
+      }
+      const shouldCallOnPress =
+        !this.longPressDetected &&
+        this.STATE !== TOUCHABLE_STATE.MOVED_OUTSIDE &&
+        this.pressOutTimeout === null;
+
+      this.handleGoToUndermined();
+      if (shouldCallOnPress) {
+        // Calls only inside component whether no long press was called previously
+        this.props.onPress && this.props.onPress();
+      }
+      this.longPressDetected = false;
+    }
+  };
+
+  onLongPressStateChange = ({ nativeEvent: { state } }) => {
+    // Handle only if state is active and finger didn't move outside component.
+    if (
+      state === State.ACTIVE &&
+      this.props.onLongPress &&
+      !this.movedOutside
+    ) {
+      this.longPressDetected = true;
+      this.props.onLongPress();
+    }
+  };
+
+  componentWillUnmount() {
+    // to prevent memory leaks
+    this.reset();
+  }
+
+  onMoveIn = () => {
+    if (this.STATE === TOUCHABLE_STATE.MOVED_OUTSIDE) {
+      // This call is not throttles with delays (like in RN's implementation).
+      this.moveToState(TOUCHABLE_STATE.BEGAN);
+    }
+  };
+
+  onMoveOut = () => {
+    // flag for longPress detection
+    this.movedOutside = true;
+    if (this.STATE === TOUCHABLE_STATE.BEGAN) {
+      this.handleMoveOutside();
+    }
+  };
+
+  // Measure component for deciding if finger is inside
+  onLayout = ({
+    nativeEvent: {
+      layout: { width, height },
+    },
+  }) => {
+    this.width = width;
+    this.height = height;
+  };
+
+  // check whether finger is inside component (including hitSlop)
+  isWithinBounds = ({ x, y }) => {
+    const { hitSlop } = this.props;
+    if (hitSlop) {
+      const { left = 0, right = 0, top = 0, bottom = 0 } = hitSlop;
+      return (
+        x > -left &&
+        x < this.width + right &&
+        y > -bottom &&
+        y < this.height + top
+      );
+    } else {
+      return x > 0 && x < this.width && y > 0 && y < this.height;
+    }
+  };
+
+  // Method for overriding
+  renderChildren(children) {
+    return children;
+  }
+
+  tap = React.createRef();
+  long = React.createRef();
+  native = React.createRef();
+  render() {
+    const extraProps = {
+      accessible: this.props.accessible !== false,
+      accessibilityLabel: this.props.accessibilityLabel,
+      accessibilityHint: this.props.accessibilityHint,
+      accessibilityComponentType: this.props.accessibilityComponentType,
+      accessibilityRole: this.props.accessibilityRole,
+      accessibilityStates: this.props.accessibilityStates,
+      accessibilityTraits: this.props.accessibilityTraits,
+      nativeID: this.props.nativeID,
+      testID: this.props.testID,
+      onLayout: this.props.onLayout,
+      hitSlop: this.props.hitSlop,
+    };
+
+    return (
+      <PanGestureHandler
```

Comment:
why do we need both pan and tap handlers? can we maybe render longpress only optionally? 99% of touchables people use don't need longpress and we could maybe only have that when longpress callback is provided as a prop

Reply:
Tap is needed because we cannot handle touch without any movement by pan. Pan is needed because it's a continuous gesture which could send event on each frame.

Reply:
Can we use GH buttons in that case? (RectButton etc)

Reply:
Not really since event delivered from this gesture handler does not tell anything about accurate position. I think it's not a big cost to use two of them and it's not worth to add extra parameters for `NativeViewGestureHandler` for this specific case.